packages/agent/src/canisterStatus/index.ts

```diff
@@ -110,10 +114,9 @@ export const request = async (options: {
           canisterId: canisterId,
         });
 
-        response.certificate;
         const lookup = (cert: Certificate, path: Path) => {
           if (path === 'subnet') {
-            const data = cert.cache_node_keys();
+            const data = fetchNodeKeys(response.certificate, agent.rootKey);
```

Comment:
`response.certificate.verify()` is done inside the  `Certificate.create()`,right? In addition to the common verification for certificates, here we need extra check of canister ranges for certificate produced by NNS subnet. 
At this point, we already know the `path === 'subnet'` thus the path canister_ranges is directly included in the  certificate tree and we can perform similar check of canister ranges as we usually do in the delegation.

Reply:
I'm deliberately moving some of this logic out of the certificate, because I don't want to overload the responsibility of the certificate class with more specific validations

Reply:
The certificate is already validated on line 119. FetchNodeKeys is just responsible for parsing the rest of the subnet metadata out of the state tree

Reply:
> The certificate is already validated on line 119, and will verify the delegation, including the canister ranges.

That's right for certificate with a delegation. I am talking about the special case of a certificate produced by the root subnet. It has no delegation and thus the canister ranges are not verified. Therefore, in addition to the common verification done to certificates, an extra check of canister ranges needs to be done when we fetch node keys in root subnet.

Reply:
We discussed it some long time ago. Here are the relevant design [slides](https://docs.google.com/presentation/d/1xbobmOhplo-OwjP41JY5MhpI7c-peyJPtNE-v_-DJKo/edit?disco=AAAA-gP82Ik).

Reply:
Here is the example.
Suppose the agent-js is making a query call to canister C on an App subnet A for the first time. Because it has no cached node keys, it sends out a read_state call requesting `/subnet` simultaneously.  However, the malicious boundary node route the read_state call and query call to a malicious node N in the root subnet. Node N returns a real certificate and a forged query response.

For the agent's perspective, the certificate is correct and it will use the node N's public key to verify the query response. The problem is that without checking the canister ranges in the certificate, it does not know that node N on the root subnet is not authorized to answer calls for canister C. That's why we request the whole `/subnet` path in addition to the node key path. In this way we can verify canister ranges directly from the tree of certificate when the delegation is absent for the root subnet.

I understand that this extra check is a special requirement and may fit in the current code perfectly. However we need it because one of feature goals is  **"subnet Y is authoritative for the canister targeted with the query"**.